### PR TITLE
Support new platform versions

### PR DESCRIPTION
**Pull request by qianz2**
An ask from Github Codespace team from email thread.

- [x] The purpose of this PR is explained in this message or in an issue. If an issue please include a reference as \#\<issue\_number\>.
- [x] Tests are included and/or updated for code changes.
- [x] Proper license headers are included in each file.

### CODE REVIEW

**Comment by qianz2** [comment]
> We might be missing updating the version for the BuildScriptGenerator similar to: https://github.com/microsoft/Oryx/pull/1323/files#diff-f9911107abdc1210d2249a15cf70159ba0b9312b614257848a88148970b4eb8dR11
> 
> May we double check please?

Updated the version files, thanks

**Comment by daniv-msft** [comment]
@qianz2 Could you please clarify the relationship between this PR and https://github.com/microsoft/Oryx/pull/1325?

**Comment by qianz2** [comment]
> @qianz2 Could you please clarify the relationship between this PR and #1325?

Yeah looks that @samruddhikhandale started working on it, @samruddhikhandale would you like to take over this work item and let us close this one?

**Comment by daniv-msft** [comment]
> > @qianz2 Could you please clarify the relationship between this PR and #1325?
> 
> Yeah looks that @samruddhikhandale started working on it, @samruddhikhandale would you like to take over this work item and let us close this one?

Just to be sure: do you need changes for your PR to be checked in? If we're almost ready, then maybe we can check that one in? As I don't expect @samruddhikhandale to contribute long term to the repo (please let me know if you think otherwise), it might be more efficient for everyone.

**Comment by qianz2** [comment]
> > > @qianz2 Could you please clarify the relationship between this PR and #1325?
> > 
> > 
> > Yeah looks that @samruddhikhandale started working on it, @samruddhikhandale would you like to take over this work item and let us close this one?
> 
> Just to be sure: do you need changes for your PR to be checked in? If we're almost ready, then maybe we can check that one in? As I don't expect @samruddhikhandale to contribute long term to the repo (please let me know if you think otherwise), it might be more efficient for everyone.

We'll need to make changes to this PR cause the some tests are failing in the build image tests, and also Java, PHP and Ruby platform binaries not getting built in the pipelines. I assume it'll be hard for @samruddhikhandale to test and trigger those pipelines from her end. We may need to avoid both working on same things. Should we close the other PR?

**Comment by daniv-msft** [comment]
> > > > @qianz2 Could you please clarify the relationship between this PR and #1325?
> > > 
> > > 
> > > Yeah looks that @samruddhikhandale started working on it, @samruddhikhandale would you like to take over this work item and let us close this one?
> > 
> > 
> > Just to be sure: do you need changes for your PR to be checked in? If we're almost ready, then maybe we can check that one in? As I don't expect @samruddhikhandale to contribute long term to the repo (please let me know if you think otherwise), it might be more efficient for everyone.
> 
> We'll need to make changes to this PR cause the some tests are failing in the build image tests, and also Java, PHP and Ruby platform binaries not getting built in the pipelines. I assume it'll be hard for @samruddhikhandale to test and trigger those pipelines from her end. We may need to avoid both working on same things. Should we close the other PR?

Sounds good to me. @samruddhikhandale, would you see any issues about closing your PR and prioritizing this one that might be easier to trigger pipelines for and complete?

File: platforms/nodejs/versionsToBuild.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@
 14.17.4
 14.17.6
 14.18.3
+14.19.1
```
Comment (samruddhikhandale) [inline platforms/nodejs/versionsToBuild.txt:19]: For updating node versions, are we missing to update the following files ? -
1. build/__nodeVersions.sh
2. src/BuildScriptGenerator/Node/NodeVersions.cs

File: platforms/python/versionsToBuild.txt [hunk 2]
```diff
@@ -11,8 +11,10 @@
 3.9.0, E3FF2839C048B25C084DEBE9B26995E310250568,
 3.9.1rc1, E3FF2839C048B25C084DEBE9B26995E310250568,
 3.9.7, E3FF2839C048B25C084DEBE9B26995E310250568,
+3.9.12, E3FF2839C048B25C084DEBE9B26995E310250568,
 
-3.10.0a2, 64E628F8D684696D,
+3.10.0a2, A035C8C19219BA821ECEA86B64E628F8D684696D,
 3.10.0, A035C8C19219BA821ECEA86B64E628F8D684696D,
+3.10.4, A035C8C19219BA821ECEA86B64E628F8D684696D,
```
Comment (samruddhikhandale) [inline platforms/python/versionsToBuild.txt:19]: For updating python versions, are we missing to update the following files ? -
1. build/__pythonVersions.sh
2. src/BuildScriptGenerator/PythonVersions.cs

File: platforms/java/versionsToBuild.txt [hunk 3]
```diff
@@ -6,4 +6,5 @@
 12.0.2, 10, 1202f536984c28d68681d51207a84b6c76e5998579132d3fe1b8085aa6a5f21e
 13.0.2, 8, 9ccc063569f19899fd08e41466f8c4cd4e05058abdb5178fa374cb365dcf5998
 14.0.2, 12, 7d5ee7e06909b8a99c0d029f512f67b092597aa5b0e78c109bd59405bbfa74fe
-17.0.1, 12, 1c0a73cbb863aad579b967316bf17673b8f98a9bb938602a140ba2e5c38f880a
\ No newline at end of file
+17.0.1, 12, 1c0a73cbb863aad579b967316bf17673b8f98a9bb938602a140ba2e5c38f880a
```
Comment (samruddhikhandale) [inline platforms/java/versionsToBuild.txt:9]: For updating java versions, are we missing to update the following files ? -
1. build/__javaVersions.sh
2. doc/runtimes/java.md
3. platforms/java/getJavaSdk.sh
4. src/BuildScriptGenerator/Java/JavaVersions.cs

File: platforms/java/versionsToBuild.txt [hunk 4]
```diff
@@ -6,4 +6,5 @@
...
+17.0.2, 8, 342e76bb9ded80aa1edd566efbae42f8f1c2366e19b9189e5941b45893b9861f
```
Comment (samruddhikhandale) [inline platforms/java/versionsToBuild.txt:10]: I could be possibly wrong, but while creating my PR I found a different SHA

> 17.0.2, 8, 13bfd976acf8803f862e82c7113fb0e9311ca5458b1decaef8a09ffd91119fa4

File: build/__rubyVersions.sh [hunk 5]
```diff
@@ -7,3 +7,5 @@ RUBY26_VERSION='2.6.6'
 RUBY26_TAR_SHA256='5db187882b7ac34016cd48d7032e197f07e4968f406b0690e20193b9b424841f'
 RUBY25_VERSION='2.5.8'
 RUBY25_TAR_SHA256='0391b2ffad3133e274469f9953ebfd0c9f7c186238968cbdeeb0651aa02a4d6d'
+RUBY31_VERSION='3.1.1'
+RUBY31_TAR_SHA256='fe6e4782de97443978ddba8ba4be38d222aa24dc3e3f02a6a8e7701c0eeb619d'
```
Comment (samruddhikhandale) [inline build/__rubyVersions.sh:11]: I could be possibly wrong, but while creating my PR I found a different SHA

RUBY31_TAR_SHA256='7aefaa6b78b076515d272ec59c4616707a54fc9f2391239737d5f10af7a16caa'

File: build/__phpVersions.sh [hunk 6]
```diff
@@ -10,6 +10,8 @@ COMPOSER1_10_VERSION='1.10.19'
 COMPOSER1_10_SETUP_SHA384='756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3'
 COMPOSER2_0_VERSION='2.0.8'
 COMPOSER2_0_SETUP_SHA384='756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3'
+COMPOSER2_2_VERSION='2.2.9'
+COMPOSER2_2_SETUP_SHA384='756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3'
 PHP81_VERSION='8.1.0'
```
Comment (samruddhikhandale) [inline build/__phpVersions.sh:15]: I believe I mentioned a version update for PHP as well to 8.1.4

Comment (samruddhikhandale) [reply build/__phpVersions.sh:15]: I see `platforms/php/versionsToBuild.txt` correctly updated, maybe we need to update here as well?

File: src/BuildScriptGenerator/PhpVersions.cs [hunk 7]
```diff
@@ -14,6 +14,9 @@ public static class PhpVersions
         public const string Composer110SetupSha384 = "756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3";
         public const string Composer20Version = "2.0.8";
         public const string Composer20SetupSha384 = "756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3";
+        public const string Composer22Version = "2.2.9";
+        public const string Composer22SetupSha384 = "756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3";
+
         public const string Php81Version = "8.1.0";
```
Comment (samruddhikhandale) [inline src/BuildScriptGenerator/PhpVersions.cs:20]: Maybe an update for `Php81Version` and `Php81Version`

File: build/constants.yaml [hunk 8]
```diff
@@ -133,6 +133,8 @@
     composer1-10-setup-sha384: 756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3
     composer2-0-version: 2.0.8
     composer2-0-setup-sha384: 756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3
+    composer2-2-version: 2.2.9
+    composer2-2-setup-sha384: 756890a4488ce9024fc62c56153228907f1545c228516cbf63f885e036d37e9a59d27d63f46af1d4d07ee0f76181c7d3
     # hashes are for .tar.xz
     php81-version: 8.1.0
```
Comment (samruddhikhandale) [inline build/constants.yaml:139]: Maybe we are missing updates for Python, PHP, Node, Java and Maven?

**Review by samruddhikhandale — COMMENTED** [review]
Comparing this PR with #1325, I found out these differences. As it was my first time creating an oryx PR, I could have made incorrect changes but wanted to point out the differences just in case it helps 😄

**Comment by samruddhikhandale** [comment]
Also, my PR had the following independent updates which I didn't mention in my email. @daniv-msft Do I create a separate PR or @qianz2 could help them with it as well? 
- Hugo 0.96.0
- Conda 4.11.0
- Yarn 1.22.18

Also, it has changes to the [vso.focal.Dockerfile](https://github.com/microsoft/Oryx/pull/1325/files#diff-b07fc2ed29986ccddfada74c20caf8cf2e8163ceb39e14ff4b636fb4242255d2) file which are codespace specific changes. These changes majorly depend on this current PR. Not sure if I should create a follow up PR?

File: platforms/java/versionsToBuild.txt (hunk 3 above)
Comment (qianz2) [reply platforms/java/versionsToBuild.txt:9]: We're supporting dynamic installation for these versions, the above files you mentioned we only update it for runtime base images. So here no need to update them. Same for other platforms.

File: build/__phpVersions.sh (hunk 6 above)
Comment (qianz2) [reply build/__phpVersions.sh:15]: Same here. These files not updated since these versions are for runtimes.

**Comment by qianz2** [comment]
> Also, my PR had the following independent updates which I didn't mention in my email. @daniv-msft Do I create a separate PR or @qianz2 could help them with it as well?
> 
> * Hugo 0.96.0
> * Conda 4.11.0
> * Yarn 1.22.18
> 
> Also, it has changes to the [vso.focal.Dockerfile](https://github.com/microsoft/Oryx/pull/1325/files#diff-b07fc2ed29986ccddfada74c20caf8cf2e8163ceb39e14ff4b636fb4242255d2) file which are codespace specific changes. These changes majorly depend on this current PR. Not sure if I should create a follow up PR?

@samruddhikhandale I would recommend us to have a separate PR since there're too many changes on a single PR, it may make us hard to check in.

**Comment by samruddhikhandale** [comment]
> @samruddhikhandale I would recommend us to have a separate PR since there're too many changes on a single PR, it may make us hard to check in.

@qianz2 I could help contribute the following changes in a separate PR. I do have repo access now, and I had already tested my changes locally last week. Let me know if I should go ahead with the new non-overlapping/non-duplicate PR? 

- Hugo 0.96.0
- Conda 4.11.0
- Yarn 1.22.18

**Comment by qianz2** [comment]
> > @samruddhikhandale I would recommend us to have a separate PR since there're too many changes on a single PR, it may make us hard to check in.
> 
> @qianz2 I could help contribute the following changes in a separate PR. I do have repo access now, and I had already tested my changes locally last week. Let me know if I should go ahead with the new non-overlapping/non-duplicate PR?
> 
> * Hugo 0.96.0
> * Conda 4.11.0
> * Yarn 1.22.18

Yeah sure, thanks @samruddhikhandale

File: platforms/java/versionsToBuild.txt (hunk 3 above)
Comment (samruddhikhandale) [reply platforms/java/versionsToBuild.txt:9]: That's good to know. Thanks!

**Review by daniv-msft — APPROVED** [review]
Looks good to me! Thanks @qianz2

**Review by daniv-msft — APPROVED** [review]

**Comment by william-msft** [comment]
@qianz2 may we add the python `3.9.12` to our testing. We can update https://github.com/microsoft/Oryx/blob/21ef7ce9cd1b58b68a8e5da304c9ed0ff6bf1d7d/tests/Oryx.BuildImage.Tests/Python/PythonDynamicInstallationTest.cs#L82
or have another similar test for our non-preview versions. Please let's also add tests for our other new supported version, coming from this PR.

**Comment by william-msft** [comment]
Let's make sure to get passing corresponding `PlatformBinary` pipeline invocations please. For example: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5972343&view=logs&j=7ac015d6-8807-5cdd-03d7-a38a0d4fbedf&t=23442349-5208-5b5b-d4fe-e5e3d2cdf805&l=69

File: platforms/python/versionsToBuild.txt [hunk 9]
```diff
@@ -11,8 +11,10 @@
 3.9.0, E3FF2839C048B25C084DEBE9B26995E310250568,
 3.9.1rc1, E3FF2839C048B25C084DEBE9B26995E310250568,
 3.9.7, E3FF2839C048B25C084DEBE9B26995E310250568,
+3.9.12, E3FF2839C048B25C084DEBE9B26995E310250568,
```
[the remark this answers is not in the document]
Comment (qianz2) [reply platforms/python/versionsToBuild.txt:14]: Thanks, will update it

**Comment by qianz2** [comment]
> Let's make sure to get passing corresponding `PlatformBinary` pipeline invocations please. For example: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5972343&view=logs&j=7ac015d6-8807-5cdd-03d7-a38a0d4fbedf&t=23442349-5208-5b5b-d4fe-e5e3d2cdf805&l=69

Yeah, the integration tests passed now as we updated several different platforms in a single PR, the tests failures caused by not having all platform versions updated in the same time. I'll trigger it again to make sure of it.

**Comment by cormacpayne** [comment]
Just a reminder that we still have the faulty Python 3.9.12 SDK in the dev storage account, so any future `PublishSdkToProd` execution will publish this version. It's still around for local testing if we want to determine what error is stemming from using the SDK, but we should remove it whenever we decide to publish new test SDKs to the dev storage account, as well as publish those SDKs to production later on.

File: platforms/python/versionsToBuild.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (william-msft) [reply platforms/python/versionsToBuild.txt:14]: Thanks Stella, `./build/generateConstants.sh` can help us populate the fields.

File: platforms/java/getJavaSdk.sh [hunk 10]
```diff
@@ -47,15 +50,14 @@ downloadJavaSdk()
     IFS='.' read -ra VERSION_PARTS <<< "$JDK_VERSION"
     majorVersion="${VERSION_PARTS[0]}"
 
-    # Version 17.0.1 has a different url format than rest of the versions, so special casing it.
-    if [ "$JDK_VERSION" == "17.0.1" ]; then
-        local buildNumber="12"
-        local url="https://download.java.net/java/GA/jdk17.0.1/2a2082e5a09d4267845be086888add4f/12/GPL/openjdk-17.0.1_linux-x64_bin.tar.gz"
+    if [ "$JDK_VERSION" == "17.0.2" ]; then
+        local url="https://download.java.net/java/GA/jdk17.0.2/dfd4a8d0985749f896bed50d7138ee7f/8/GPL/openjdk-17.0.2_linux-x64_bin.tar.gz"
```
Comment (arroyc) [inline platforms/java/getJavaSdk.sh:54]: nit: constantify this

**Review by arroyc — COMMENTED** [review]
nit: add some new version tests

**Comment by samruddhikhandale** [comment]
When do we plan on fixing the tests failures? Looks like uploading new platform versions (in my case NodeJS) is also causing my oryx-validation build to fail for branch [samruddhikhandale/version-update-dockerfile](https://github.com/microsoft/Oryx/tree/samruddhikhandale/version-update-dockerfile)

**Comment by qianz2** [comment]
> When do we plan on fixing the tests failures? Looks like uploading new platform versions (in my case NodeJS) is also causing my oryx-validation build to fail for branch [samruddhikhandale/version-update-dockerfile](https://github.com/microsoft/Oryx/tree/samruddhikhandale/version-update-dockerfile)

We have a fix PR: https://github.com/microsoft/Oryx/pull/1337/files, which will be checked in soon to fix validation failures. It'll be updated here in this PR too.

**Review by daniv-msft — DISMISSED** [review]

File: doc/supportedPlatformVerions.md [hunk 11]
```diff
@@ -238,16 +242,19 @@
 - 3.9.0b1
 - 3.9.0
 - 3.9.1rc1
+- 3.9.12
```
Comment (daniv-msft) [inline doc/supportedPlatformVerions.md:245]: As discussed separately, @qianz2 will send a separate PR to remove these last references to 3.9.12. We don't want to trigger a full rebuild.

File: build/__javaVersions.sh [hunk 12]
```diff
@@ -1,4 +1,4 @@
 # This file was auto-generated from 'constants.yaml'. Changes may be overridden.
```
Comment (cormacpayne) [inline build/__javaVersions.sh:1]: For my own personal inventory, this PR is adding the following platform versions:

- Java 17.0.2
- Node 16.14.2
- PHP 7.4.28 and 8.1.4
- Ruby 3.0.0 and 3.1.1

I see some versions are added to buster and focal as well (such as Node 14.19.1) -- just curious if these versions were previously meant to be added to buster and focal and weren't, or if there's a separate reason we're adding them now.

File: platforms/python/versionsToBuild.txt [hunk 13]
```diff
@@ -12,7 +12,8 @@
 3.9.1rc1, E3FF2839C048B25C084DEBE9B26995E310250568,
 3.9.7, E3FF2839C048B25C084DEBE9B26995E310250568,
 
-3.10.0a2, 64E628F8D684696D,
```
Comment (cormacpayne) [inline platforms/python/versionsToBuild.txt:15]: Do we need to revert the changes in this file since we aren't adding or modifying Python versions, or is this a separate change?

**Review by cormacpayne — APPROVED** [review]

**Comment by cormacpayne** [comment]
Also would be great if we could double-check the platform binaries that are published to the development storage account before we merge this PR and publish binaries to production, just to ensure that everything lines up with what we expect.

File: tests/Oryx.BuildImage.Tests/Java/JavaDynamicInstallationTest.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Oryx.BuildScriptGenerator.Common;
+using Microsoft.Oryx.BuildScriptGenerator.Java;
+using Microsoft.Oryx.Tests.Common;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Oryx.BuildImage.Tests
```
Comment (daniv-msft) [inline tests/Oryx.BuildImage.Tests/Java/JavaDynamicInstallationTest.cs:14]: I couldn't find where these tests are executed in the pipeline. Looking at the two "Publish test results" steps for both Build and Runtime, I couldn't find them (or the previously existing tests for that matter). Do you know where they are executed?

File: build/__javaVersions.sh (hunk 12 above)
Comment (qianz2) [reply build/__javaVersions.sh:1]: These versions are auto added to stretch, buster and focal when we triggered the platform binary pipelines. For each SDK we have these three kinds so that users can use it from different types of images.

File: platforms/python/versionsToBuild.txt (hunk 13 above)
Comment (qianz2) [reply platforms/python/versionsToBuild.txt:15]: The sha for 3.10.0a2 isn't correct so I updated it here. And since we've already had these versions in both dev and prod, it won't add or modify the existing ones.

File: tests/Oryx.BuildImage.Tests/Java/JavaDynamicInstallationTest.cs (hunk 14 above)
Comment (qianz2) [reply tests/Oryx.BuildImage.Tests/Java/JavaDynamicInstallationTest.cs:14]: Yes, they're executed, the total number of tests is 372(scroll down at the bottom of build image tests), compared to previously we had 367.

Comment (daniv-msft) [reply tests/Oryx.BuildImage.Tests/Java/JavaDynamicInstallationTest.cs:14]: Thanks!

**Review by daniv-msft — APPROVED** [review]
